ITEP-68935 Feature: dynamic camera configuration in K8S (1st iteration) (#409)
open-edge-platform/scenescape: docs/user-guide/Other Topics/How-to-configure-dlstreamer-video-pipeline.md

### Generating a Pipeline Preview

The camera calibration page provides an automated pipeline generation feature:

1. **Fill in Required Fields**: enter the necessary camera configuration parameters:
   - Set the **Camera (Video Source)** (e.g., `rtsp://camera-ip:554/stream`).
   - Configure **Camera Chain** settings, if needed.
   - Select the appropriate **Model Config**.

2. **Generate Pipeline Preview**: click the **"Generate Pipeline Preview"** button.
   - The system will automatically generate a GStreamer pipeline based on your configuration.
   - The generated pipeline appears in the **Camera Pipeline** text area.
   - You can review the pipeline structure and elements.

3. **Review Generated Pipeline**: the generated pipeline will include:
   - Video source configuration based on your Camera (Video Source) field.
   - AI model integration using the selected Model Config.
   - Camera intrinsics and distortion correction, if configured.
   - Metadata publishing for SceneScape integration.

### Customizing the Generated Pipeline

After generating a pipeline preview, you can make manual adjustments:

> 1. **Edit Pipeline String**: modify the generated pipeline in the Camera Pipeline text area.

Reviewer: Do we show a warning when editing Preview directly that directly editing the preview will leave the components fields and gst pipeline string out of sync.

#### Core Pipeline Fields

- **Camera (Video Source)**: Specifies the video source command. Supported formats:
  - RTSP streams: `rtsp://camera-ip:554/stream` (raw H.264)
  - HTTP/HTTPS streams: `http://camera-ip/mjpeg` (MJPEG)
  - File sources: `file://video.ts` (relative to video folder)
- **Camera Chain**: Defines the sequence or combination of AI models to chain together in the pipeline using their short identifiers (e.g., "retail"). Models can be chained serially (one after another) or in parallel arrangements. These identifiers are defined in the model configuration file with their detailed parameters needed for pipeline generation.

> **Note**: Currently only limited model chaining is suported. See the limitations section below.

- **Camera Pipeline**: The generated or custom GStreamer pipeline string

#### Advanced Configuration

- **Decode Device**: Video decoding device settings (CPU or GPU)
- **Model Config**: References a model configuration file. Model configuration files are managed in the Models page and stored in the folder `Models/models/model_configs`. The user can upload custom model configuration files or modify existing ones using the Models page.

> > **Note**: The Model Config field references configuration files that define AI model parameters and processing settings. For details on the model configuration file format, see [Model Configuration File Format](Model-configuration-file-format.md).

Reviewer 2: ```suggestion
> **Note**: The Model Config field references configuration files that define AI model parameters and processing settings. See [Model Configuration File Format](Model-configuration-file-format.md) for more details.
```

#### Camera Intrinsics and Distortion

- **Intrinsics**: Camera focal lengths (fx, fy) and principal point coordinates (cx, cy)
- **Distortion Coefficients**: k1, k2, k3, p1, p2 for lens distortion correction

### Generating a Pipeline Preview

The camera calibration page provides an automated pipeline generation feature:

1. **Fill in Required Fields**: Enter the necessary camera configuration parameters:
   - Set the **Camera (Video Source)** field with your video source (e.g., `rtsp://camera-ip:554/stream`)
   - Configure **Camera Chain** settings if needed
   - Select appropriate **Model Config** from available options

2. **Generate Pipeline Preview**: Click the **"Generate Pipeline Preview"** button
   - The system will automatically generate a GStreamer pipeline based on your configuration
   - The generated pipeline appears in the **Camera Pipeline** text area
   - You can review the pipeline structure and elements

> 3. **Review Generated Pipeline**: The generated pipeline will include:
>    - Video source configuration based on your Camera (Video Source) field
>    - AI model integration using the selected Model Config
>    - Camera intrinsics and distortion correction if configured
>    - Metadata publishing for SceneScape integration

Reviewer 2: ```suggestion
3. **Review Generated Pipeline**: the generated pipeline will include:
   - Video source configuration based on your Camera (Video Source) field.
   - AI model integration using the selected Model Config.
   - Camera intrinsics and distortion correction, if configured.
   - Metadata publishing for SceneScape integration.
```

### Customizing the Generated Pipeline

After generating a pipeline preview, you can make manual adjustments:

> 1. **Edit Pipeline String**: Modify the generated pipeline in the Camera Pipeline text area
>    - Add or remove GStreamer elements as needed
>    - Adjust element parameters for specific requirements
>    - Ensure the pipeline maintains compatibility with SceneScape - do not modify `gvapython` or `cameraundistort` elements

Reviewer 2: ```suggestion
1. **Edit Pipeline String**: modify the generated pipeline in the Camera Pipeline text area.
   - Add or remove GStreamer elements as needed.
   - Adjust element parameters for specific requirements.
   - Ensure the pipeline maintains compatibility with SceneScape - do not modify `gvapython` or `cameraundistort` elements.
```

> 2. **Common Customizations**:
>    - **Video Source**: Change input source type (file, RTSP, USB)
>    - **Model Parameters**: Fine-tune AI model inference settings either in model config file or the **Camera Pipeline** field

Reviewer 2: ```suggestion
2. **Common Customizations**:
   - **Video Source**: change input source type (file, RTSP, USB).
   - **Model Parameters**: fine-tune AI model inference settings either in model config file or the **Camera Pipeline** field.
```

> 3. **Validation**: The system performs preliminary checks of the pipeline and reports an error if pipeline generation is not possible, when you save the configuration or generate the pipeline preview. However, it does not validate pipeline correctness in terms of GStreamer pipeline syntax and its functionality. It is up to the user to verify that the pipeline performs as expected.

Reviewer 2: ```suggestion
3. **Validation**: when you save the configuration or generate the pipeline preview, the system performs preliminary checks of the pipeline and reports an error if pipeline generation is not possible. However, it does not validate pipeline correctness in terms of GStreamer pipeline syntax and its functionality. You need to verify that the pipeline performs as expected.
```

### Saving and Applying Configuration

> 1. **Save Camera Configuration**: Click **"Save Camera"** to apply your pipeline configuration
>    - The system automatically generates the camera pipeline if the field is empty
>    - Configuration is stored and deployed to the Kubernetes cluster
>    - The camera deployment is updated with the new pipeline

Reviewer 2: ```suggestion
1. **Save Camera Configuration**: click **"Save Camera"** to apply your pipeline configuration.
   - The system automatically generates the camera pipeline if the field is empty.
   - Configuration is stored and deployed to the Kubernetes cluster.
   - The camera deployment is updated with the new pipeline.
```

> 2. **Automatic Pipeline Generation**: If you save the form with an empty Camera Pipeline field:
>    - The system automatically generates a pipeline based on other form fields
>    - This ensures every camera has a valid pipeline configuration
>    - The generated pipeline follows SceneScape best practices and standards

Reviewer 2: ```suggestion
2. **Automatic Pipeline Generation**: if you save the form with an empty Camera Pipeline field, the system automatically generates a pipeline based on other form fields, following best practices and standards for SceneScape. This ensures every camera has a valid pipeline configuration.
```

> 3. **Error Handling**: If pipeline generation fails:
>    - Error messages are displayed in the form
>    - The form remains open for corrections
>    - Common issues include missing model configurations or invalid command syntax

Reviewer 2: ```suggestion
3. **Error Handling**: If pipeline generation fails, the form remains open for correction and error messages are displayed. Common issues include missing model configurations or invalid command syntax.
```

### Best Practices

> - **Start with Generated Pipeline**: Use the "Generate Pipeline Preview" button to create a baseline configuration
> - **Test Incrementally**: Make small changes and test each modification
> - **Validate Model Config**: Ensure your selected Model Config file exists and is properly formatted
> - **Monitor Performance**: Check camera performance after applying pipeline changes
> - **Backup Configurations**: Save working pipeline configurations for future reference

Reviewer 2: ```suggestion
- **Start with Generated Pipeline**: use the "Generate Pipeline Preview" button to create a baseline configuration.
- **Test Incrementally**: make small changes and test each modification.
- **Validate Model Config**: ensure your selected Model Config file exists and is properly formatted.
- **Monitor Performance**: check camera performance after applying pipeline changes.
- **Backup Configurations**: save working pipeline configurations for future reference.
```

### Pipeline Generation Limitations

> - Multiple model chaining is not supported yet. Only a single detection model can be used as **Camera Chain**
> - The only **Decode Device** supported for now is `CPU`
> - Distortion correction is temporarily disabled due to a bug in DLStreamer-Pipeline-Server
> - Explicit frame rate and resolution configuration is not available yet.
> - Network instability and camera disconnects are not handled gracefully for network-based (RTSP/HTTP/HTTPS) streams and may cause the pipeline to fail

Reviewer 2: ```suggestion
- Multiple model chaining is not supported yet. Only a single detection model can be used as **Camera Chain**.
- The only **Decode Device** supported for now is `CPU`.
- Distortion correction is temporarily disabled due to a bug in DLStreamer-Pipeline-Server.
- Explicit frame rate and resolution configuration is not available yet.
- Network instability and camera disconnects are not handled gracefully for network-based streams (RTSP/HTTP/HTTPS) and may cause the pipeline to fail.
```

### Troubleshooting

> - **Pipeline Generation Errors**: Check that all required fields are filled correctly
> - **Model Config Issues**: Verify the model configuration file exists in the Models page and model(s) used in the **Camera Chain** field are defined in the model config file
> - **Video Source Problems**: Ensure the Camera (Video Source) field contains a valid video source URL or device path
> - **Deployment Failures**: Check Kubernetes logs for detailed error information

Reviewer 2: ```suggestion
- **Pipeline Generation Errors**: check that all required fields are filled correctly.
- **Model Config Issues**: verify the model configuration file exists in the Models page and model(s) used in the **Camera Chain** field are defined in the model config file.
- **Video Source Problems**: ensure the Camera (Video Source) field contains a valid video source URL or a device path.
- **Deployment Failures**: check Kubernetes logs for detailed error information.
```
